### PR TITLE
Implement key-value store

### PR DESCRIPTION
**Pull request by caspervonb**
Implements key-value store as per the ADR.

### CODE REVIEW

**Review by Jarema — CHANGES_REQUESTED** [review]
Overall note: code shows that it's port from Go client and can be optimized/cleared out,
though it seems that it's easiest way to get features of KV right, coverage them with tests and then optimise.

File: src/jetstream_kv.rs [hunk 1]
```diff
@@ -0,0 +1,909 @@
...
+                    Some(entry.value)
+                } else {
+                    None
+                };
+
+                Ok(value)
+            }
+            Ok(None) => Ok(None),
+            Err(err) => Err(err),
+        }
```
Comment (Jarema) [inline src/jetstream_kv.rs:409]: Little hard to read.
Something like that is IMHO clearer to read:

```suggestion
        match self.entry(key) {
            Ok(Some(entry)) => {
                match entry.operation {
                    KeyValueOperation::Put => Ok(Some(entry.value)),
                    _ => Ok(None),
                }
            }
            Ok(None) => Ok(None),
            Err(err) => Err(err),
        }
```

File: src/jetstream_kv.rs [hunk 2]
```diff
@@ -0,0 +1,909 @@
...
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use nats::jetstream::KeyValueConfig;
+    /// # fn main() -> std::io::Result<()> {
+    /// # let client = nats::connect("demo.nats.io")?;
+    /// # let context = nats::jetstream::new(client);
+    /// #
+    /// let bucket = context.create_key_value(&KeyValueConfig {
```
[the remark this answers is not in the document]
Comment (caspervonb) [reply src/jetstream_kv.rs:172]: Currently we don't have builders for any of the other types of configurations.

File: src/jetstream_kv.rs [hunk 3]
```diff
@@ -0,0 +1,909 @@
...
+    /// })?;
+    ///
+    /// let key_value = context.key_value("key_value")?;
+    ///
+    /// # context.delete_key_value("key_value")?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn key_value(&self, bucket: &str) -> io::Result<KeyValueStore> {
+        if !self.connection.is_server_compatible_version(2, 6, 2) {
```
[the remark this answers is not in the document]
Comment (caspervonb) [reply src/jetstream_kv.rs:128]: Only checked twice, so teetering on the edge extract vs repeat imo.

File: src/jetstream_kv.rs [hunk 4]
```diff
@@ -0,0 +1,909 @@
...
+    Put,
+    /// A value was deleted from a bucket
+    Delete,
+    /// A value was purged from a bucket
+    Purge,
+}
+
+// Helper to extract key value operation from message headers
+fn kv_operation_from_maybe_headers(maybe_headers: Option<&Headers>) -> KeyValueOperation {
+    if let Some(headers) = maybe_headers.as_ref() {
```
[the remark this answers is not in the document]
Comment (caspervonb) [reply src/jetstream_kv.rs]: Yep!

File: src/jetstream_kv.rs (hunk 1 above)
Comment (caspervonb) [reply src/jetstream_kv.rs]: Agreed and done.

File: src/jetstream_kv.rs [hunk 5]
```diff
@@ -0,0 +1,906 @@
...
+    /// #
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn create(&self, key: &str, value: impl AsRef<[u8]>) -> io::Result<u64> {
+        if !is_valid_key(key) {
+            return Err(io::Error::new(io::ErrorKind::InvalidInput, "invalid key"));
+        }
+
+        let revision = if let Ok(Some(entry)) = self.entry(key) {
```
Comment (Jarema) [inline src/jetstream_kv.rs:469]: Setting revision to `0` if `self.entry(key)` is Err.

I guess it was not intentional and was meant only for `Some`, right?

File: src/jetstream_kv.rs [hunk 6]
```diff
@@ -0,0 +1,906 @@
...
+        subject.push_str(&self.prefix);
+        subject.push_str(ALL_KEYS);
+
+        let subscription = self.context.subscribe_with_options(
+            &subject,
+            &SubscribeOptions::ordered()
+                .headers_only()
+                .deliver_last_per_subject()
+                .enable_flow_control()
+                .idle_heartbeat(Duration::from_millis(5000)),
```
Comment (Jarema) [inline src/jetstream_kv.rs:658]: That heartbeat value is consistent across many methods. Should be refactored into const.

File: src/jetstream_kv.rs (hunk 5 above)
Comment (caspervonb) [reply src/jetstream_kv.rs:469]: That's intentional, I'll flatten this flow a bit to be more along the happy path.

File: src/jetstream_kv.rs (hunk 6 above)
Comment (caspervonb) [reply src/jetstream_kv.rs:658]: Yeah it's also arbitrary, think I derived the value from the JavaScript implementation but lets just have the `subscribe` call default it.

**Comment by caspervonb** [comment]
Test is unexpectedly flaky on CI, CI is probably right however so digging into that further first thing in the morning.

**Comment by derekcollison** [comment]
Possible to squash down and close all resolved issues to ease review?

**Review by derekcollison — APPROVED** [review]
Did not dive crazy deep but in general LGTM - Thanks!

**Comment by Jarema** [comment]
API looks ok! 
I'm making a deeper dive but in the meantime it would be nice to fix lint issues.

**Comment by Jarema** [comment]
LGTM.

**Review by Jarema — APPROVED** [review]

**Pull request merged**